# 1.4 데이터 핸들링 - 판다스

- 넘파이는 저수준 API가 대부분인 반면, 판다스는 내부 데이터 뿐만 아니라 CSV 등의 파일을 쉽게 DataFrame으로 변경해 분석을 수행할 수 있게 해준다.

- Series는 칼럼이 하나뿐인 데이터 구조체이고, DataFrame은 칼럼이 여러 개인 데이터 구조체이다.
- 즉 DataFrame은 여러 개의 Series로 이루어졌다고 할 수 있다.

## 1.4.1 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API

In [1]:
import pandas as pd

www.kaggle.com 에서 데이터 csv 파일을 다운받을 수 있다. 
타이타닉 탑승자 데이터 파일을 사용할 것이다.

In [2]:
titanic_df = pd.read_csv('titanic_train.csv') # 파이썬 파일과 동일한 디렉토리에 csv 파일을 위치시키고, data frame에 담아주김
print('titanic 변수 type:',type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [3]:
titanic_df.head(3) # .head(n)으로 data frame에서 n개의 데이터만을 미리보기

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [4]:
print('DataFrame 크기: ', titanic_df.shape) # .shape으로 data frame의 크기 확인

DataFrame 크기:  (891, 12)


-> 총 891개의 데이터가 있고, 하나당 12개의 정보들을 갖고 있는 것을 알 수 있다.(891개의 row와 12개의 column)

In [5]:
titanic_df.info() # .info()로 데이터의 타입, null 데이터의 개수 등 확인

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
titanic_df.describe() # .describe()으로 숫자형 column에 대한 대략적인 '분포도' 확인

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- **count** : Not Null인 데이터 건수
- **mean** : 전체 데이터의 평균 값
- **std** : 표준편차
- **min** : 최솟값
- **max** : 최댓값
- **25%, 50%, 75%** : 각 위치에 해당하는 percentile 값

-> 'Survived' 칼럼의 경우에는 min값과 max값이 0 아니면 1 이므로, 0과 1로 이루어진 **카테고리 칼럼**이라고 추측할 수 있다. 
<br>
-> 마찬가지로 'Pclass' 칼럼의 경우에도 1, 2, 3으로만 이루어져있는 카테고리 칼럼이라고 볼 수 있다.

In [7]:
value_counts = titanic_df['Pclass'].value_counts() # .value_counts()로 데이터값 건수 확인(횟수의 내림차순으로 출력됨)
print(value_counts)

Pclass
3    491
1    216
2    184
Name: count, dtype: int64


In [8]:
titanic_pclass = titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환한다.

In [9]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

In [10]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


In [11]:
print('titanic_df 데이터 건수: ', titanic_df.shape[0])
print('기본 설정인 dropna=True로 value_counts()')
# value_counts()는 디폴트로 dropna=True이므로 value_counts(dropna=True)와 동일.
print(titanic_df['Embarked'].value_counts())
print(titanic_df['Embarked'].value_counts(dropna=False))

titanic_df 데이터 건수:  891
기본 설정인 dropna=True로 value_counts()
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


'dropna=False' 옵션을 인자로 전달하면 NaN 값을 따로 빼서 'value_counts'를 할 수 있다.

## 1.4.2 DataFrame과 리스트, 딕셔너리, 넘파이 nadarry 상호 변환

### 넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기

In [12]:
import numpy as np

col_name1=['col1']
list1 = [1, 2, 3]
array1 = np.array(list1)

print('array1 shape:', array1.shape )
df_list1 = pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)
df_array1 = pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


In [13]:
# 3개의 컬럼명이 필요함. 
col_name2=['col1', 'col2', 'col3']

# 2행x3열 형태의 리스트와 ndarray 생성 한 뒤 이를 DataFrame으로 변환. 
list2 = [[1, 2, 3],
         [11, 12, 13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape )
df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2 = pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


In [14]:
# Key는 컬럼명으로 매핑, Value는 리스트 형(또는 ndarray)
dict = {'col1':[1, 11], 'col2':[2, 22], 'col3':[3, 33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


### DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용하므로, DataFrame을 이용하더라도 다시 ndarray로 변환해야 하는 경우가 빈번하게 발생한다.

In [15]:
# DataFrame을 ndarray로 변환
array3 = df_dict.values # .values로 dataframe의 값들을 넣어주기

print('df_dict.values 타입:', type(array3), 'df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [16]:
# DataFrame을 리스트로 변환
list3 = df_dict.values.tolist() # .values와 .tolist()로 dataframe의 값들을 list 형으로 변환
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

# DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입:', type(dict3))
print(dict3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


## 1.4.3 DataFrame의 칼럼 데이터 세트 생성과 수정

In [17]:
titanic_df['Age_0']=0 # 새로운 칼럼을 추가, 모든 값들이 0으로 채워짐
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

titanic_df['Age_0']=0과 같이 Series에 상숫값을 할당하면 모든 데이터세트에 **일괄적으로 적용**된다.

In [18]:
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

In [19]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

기존의 값도 일괄적으로 변경할 수 있다.

## 1.4.4 DataFrame 데이터 삭제

DataFrame에서 데이터의 삭제는 drop() 메서드를 사용한다. <br><br>
drop() 메서드의 원형은 다음과 같다. <br>
`DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors'raise')`

이 중 가장 중요한 파라미터는 labels, axis, inplace이다. <br><br>

axis 0은 row 방향 축, axis 1은 column 방향 축이다. 그러므로 **labels에 원하는 칼럼명을 입력하고 axis=1을 입력하면 지정된 칼럼을 드롭한다.**

<br>
drop()를 쓰는 대부분의 경우는 칼럼을 드랍하는 경우이다.

In [20]:
titanic_drop_df = titanic_df.drop('Age_0', axis=1 )
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

'Age_0' 칼럼이 삭제된 것을 볼 수 있다.

In [21]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

기존의 `titanic_df` data frame을 다시 보면, 위에서 삭제했던 'Age_0' 칼럼이 남아있는 것을 볼 수 있다. 
<br>
이는 **'inplace = false'** 로 디폴트로 설정되었기 때문이다. 
<br>

'inplace=False'이면 자기 자신의 데이터는 삭제하지 않고, 반환하는 것만 삭제된 결과의 dataframe을 반환한다.
<br>
'inplace=True'으로 설정하면 자신 자체의 데이터를 삭제한다. 이때 따로 dataframe을 반환하지 않기 때문에, 기존 dataframe에 담아 None값이 되는 일이 없도록 주의해야 한다.


또한 다음과 같이 여러 개의 칼럼을 삭제할 수도 있다.

In [22]:
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print(' inplace=True 로 drop 후 반환된 값:',drop_result)
titanic_df.head(3)

 inplace=True 로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

다음과 같이 row를 삭제할 수도 있다.

In [23]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))

titanic_df.drop([0,1,2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


**[drop() 메서드 사용 방법 정리]**
<br>
- axis : DataFrame의 row를 삭제할 때는 axis=0, column을 삭제할 때는 axis=1로 설정
- 원본은 유지하고 드롭된 dataframe을 새로운 객체 변수로 받고 싶다면 **inplace=True**로 설정(디폴트 값은 False)
- 원본에 드롭된 결과를 적용할 때는 **inplace=True**를 적용

## 1.4.5 Index 객체

`(data frame 이름).index`로 Index 객체를 추출할 수 있다.

In [24]:
# 원본 파일 재 로딩 
titanic_df = pd.read_csv('titanic_train.csv')
# Index 객체 추출
indexes = titanic_df.index
print(indexes)
# Index 객체를 실제 값 arrray로 변환 
print('Index 객체 array값:\n',indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

Index 객체는 식별성 데이터를 1차원 array로 가지고 있으며, ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능하다.

In [27]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


하지만 ndarray와의 차이점은, **한 번 만들어진 Index 객체는 변경이 불가하다는 것**이다.

In [26]:
indexes[0] = 5

TypeError: Index does not support mutable operations

변경하려고 하면 위와 같이 오류가 뜨는 것을 볼 수 있다.

또한 Series 객체에 연산 함수를 적용할 때 **Index는 연산에서 제외된다.**
<br>
**Index는 오직 '식별용'으로만 사용된다.**

In [32]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n',(series_fair + 3).head(3) )

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


`reset_index()` 메서드는 새롭게 인덱스를 연속 숫자 형으로 할당하며, 기존 인덱스는 'index'라는 새로운 칼럼명으로 추가한다.
<br>
이는 인덱스가 연속된 int형 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용한다.

In [34]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

In [35]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:',type(value_counts))

new_value_counts = value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:',type(new_value_counts))

### before reset_index ###
Pclass
3    491
1    216
2    184
Name: count, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   Pclass  count
0       3    491
1       1    216
2       2    184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


Series에 `reset_index()`를 적용하면 새롭게 연속 숫자형 인덱스가 만들어지고, 기존 인덱스는 'index' 칼럼명으로 추가되면서 **데이터 타입이 DataFrame으로 변환됨**을 알 수 있다.
<br>
이때 파라미터를 `drop=True`로 설정하면 **기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제된다.** (새로운 칼럼이 추가되지 않으므로 그대로 Series로 유지된다.)

## 1.4.5 데이터 셀렉션 및 필터링

판다스의 경우에는 '[]'연산자로 단일 값 추출, 슬라이싱 등의 데이터 셀렉션을 수행한 반면, 판다스의 경우에는 `iloc[]`, `loc[]` 연산자를 통해 동일한 작업을 수행한다.

### DataFrame의 [ ] 연산자

넘파이에서 [] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있었다.
<br>
하지만 DataFrame 바로 뒤에 있는 '[]'안에 들어갈 수 있는 것은 칼럼명 문자, 또는 인덱스로 변환 가능한 표현식이다. 

In [30]:
print('단일 컬럼 데이터 추출:\n', titanic_df[ 'Pclass' ].head(3))
print('\n여러 컬럼들의 데이터 추출:\n', titanic_df[ ['Survived', 'Pclass'] ].head(3))
print('[ ] 안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 컬럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 컬럼들의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

[] 안에 ''로 감싸진 문자값이 아니라 숫자형 데이터가 들어갔기 때문에 마지막 줄에서 오류가 발생한다.

In [36]:
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

하지만 위에서처럼 슬라이싱을 이용하면 원하는 결과를 반환한다.**(숫자 자체는 안 되지만, 슬라이싱은 가능)**

In [ ]:
titanic_df[ titanic_df['Pclass'] == 3].head(3)

불린 인덱싱도 가능하다.(자주 사용된다.)

#### [ [ ] 연산자의 혼돈을 방지하기 위한 가이드 ] 
- DataFrame 바로 뒤의 [] 연산자는 넘파이의 []나 Series의 []와 다르다.
- DataFrame 바로 뒤의 [] 내 입력값은 칼럼명(또는 칼럼명의 리스트)을 지정해 칼럼 지정 연산에 사용하거나, 불린 인덱스 용도로만 사용해야 한다.
- 슬라이싱으로 데이터를 추출하는 방법은 사용하지 않는 것이 좋다.

### DataFrame iloc[ ] 연산자

판다스는 DataFrame의 로우나 컬럼을 지정하여 데이터를 선택할 수 있는 인덱싱 방식으로 `iloc[]`와 `loc[]`를 제공한다.
<br>
`iloc[]`은 위치 기반 인덱싱 방식이고, `loc[]`는 명칭 기반 인덱싱 방식으로 동작한다.
<br>
위치 기반 인덱싱은 행과 열 위치를, 0을 출발점으로 하는 세로축, 가로축 좌표 정수값으로 지정하는 방식이다.
<br>
명칭 기반 인덱싱은 데이터 프레임의 인덱스 값으로 행 위치를, 칼럼의 명칭으로 열 위치를 지정하는 방식이다.

In [37]:
data = {'Name': ['Chulmin', 'Eunkyung','Jinwoong','Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']
       }
data_df = pd.DataFrame(data, index=['one','two','three','four'])
data_df

Name  Year  Gender
one     Chulmin  2011    Male
two    Eunkyung  2016  Female
three  Jinwoong  2015    Male
four    Soobeom  2015    Male

In [39]:
data_df.iloc[0, 0]

'Chulmin'

In [40]:
# 아래 코드는 오류를 발생합니다. 
data_df_reset.iloc[0,'Name']

NameError: name 'data_df_reset' is not defined

`iloc[]`의 열 위치에 위치 정숫값이 아닌 칼럼 명칭을 입력하여 오류가 발생한 것을 볼 수 있다.

In [45]:
# 아래 코드는 오류를 발생합니다.
data_df.iloc['one', 0]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

`iloc[]`은 **열 위치에 -1을 입력**하여 가장 마지막 열 데이터를 가져오는데 자주 사용된다.
<br>
특히 머신 러닝 학습 데이터의 **맨 마지막 칼럼이 타깃 값인 경우**가 많은데, 이 경우 `iloc[:, -1]`을 하면 맨 마지막 칼럼의 값, 즉 **타깃값**을 가져오고, `iloc[:, :-1]`을 하게 되면 처음부터 맨 마지막 칼럼을 제외한 모든 칼럼의 값, 즉 **피처값**들을 가져오게 된다.

In [43]:
print("\n 맨 마지막 칼럼 데이터 [:, -1] \n", data_df.iloc[:, -1])
print("\n 맨 마지막 칼럼을 제외한 모든 데이터 [:, :-1] \n", data_df.iloc[:, :-1])


 맨 마지막 칼럼 데이터 [:, -1] 
 one        Male
two      Female
three      Male
four       Male
Name: Gender, dtype: object

 맨 마지막 칼럼을 제외한 모든 데이터 [:, :-1] 
            Name  Year
one     Chulmin  2011
two    Eunkyung  2016
three  Jinwoong  2015
four    Soobeom  2015


반면 `loc[]`에서는 반대로 칼럼 명칭으로 값을 추출할 수 있다.

In [44]:
data_df.loc['one', 'Name']

'Chulmin'

Index값이 왜 칼럼 명칭이냐고 생각할 수도 있지만, 행 위치를 나타내는 고유의 '명칭' 정도로 생각할 수 있겠다.

In [48]:
# 다음 코드는 오류를 발생시킵니다.
data_df.loc[0, 'Name']

KeyError: 0

하지만 그렇다고 행 위치에 0을 숫자 그대로 입력하면 오류가 발생한다. 

`loc[]`에 슬라이싱 기호 ':'를 적용할 때 유의할 점이 있다. 
<br>
일반적으로는 슬라이싱을 할 때 (종료값-1)까지 범위가 지정되지만, `loc[]`에 슬라이싱 기호를 적용하면 **종료값까지 포함**하게 된다.
<br>
이는 명칭 기반 인덱싱의 특성 때문이다. 명칭은 숫자형이 아닐 수 있기 때문에 -1을 할 수가 없다.

In [49]:
print('위치기반 iloc slicing\n', data_df.iloc[0:1, 0],'\n')
print('명칭기반 iloc slicing\n', data_df.loc['one':'two', 'Name'])

위치기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭기반 iloc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


위치 기반 인덱싱에서는 1개의 행을 반환했지만, 명칭 기반 인덱싱에서는 2개의 행이 반환된 것을 볼 수 있다. 즉 **똑같이 수를 입력해도 명칭 기반 인덱싱에서 추출되는 데이터의 개수가 1개 더 많다.**

## boolean indexing

실전에서는 불린 인덱싱으로 데이터를 가져오는 경우가 가장 많다. 처음부터 가져올 값을 조건으로 []내에 입력하면 자동으로 원하는 값을 필터링하기 때문이다.

In [75]:
# 여기부터 주석 달음!!
titanic_df = pd.read_csv('titanic_train.csv') #타이타닉 train set을 담은 data frame 생성
titanic_boolean = titanic_df[titanic_df['Age']>60] #age가 60 이상인 것만 담음
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

In [76]:
titanic_df[titanic_df['Age'] > 60][['Name','Age']].head(3) # age가 60 이상인 것들의 'name', 'age'만 미리보기(컬럼이 두 개 이상일 때는 [[]] 사용!!)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [77]:
titanic_df.loc[titanic_df['Age'] > 60, ['Name','Age']].head(3) # loc 사용해서 동일하게 미리보기

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

In [53]:
titanic_df[ (titanic_df['Age'] > 60) & (titanic_df['Pclass']==1) & (titanic_df['Sex']=='female')]
# age가 60 초과이고, Pclass가 1이고, Sex가 female인 데이터만 나타내기

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

위와 같이 복합 조건 연산자를 사용하여 인덱싱 할 수도 있다.

In [54]:
cond1 = titanic_df['Age'] > 60
cond2 = titanic_df['Pclass']==1
cond3 = titanic_df['Sex']=='female'
titanic_df[ cond1 & cond2 & cond3]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

또는 이런 식으로 개별 조건을 변수에 할당하여 불린 인덱싱을 수행할 수도 있다.

## 1.4.6 정렬, Aggregation 함수, GroupBy 적용

### DataFrame, Series의 정렬 - sort_values()

DataFrame과 Series의 정렬을 위해서는 `sort_values()` 메서드를 이용한다. 
<br>
`sort_values()`의 주요 입력 파라미터는 **by, ascending, inplace**이다. 
- **by** : 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행한다.
- **ascending** : 디폴트 값인 True는 오름차순, False는 내림차순으로 정렬한다.
- **inplace** : 디폴트 값인 False는 기존 데이터를 유지하고 정렬된 데이터를 반환하는 반면, True는 기존 데이터에 정렬된 결과를 그대로 적용한다.

In [78]:
titanic_sorted = titanic_df.sort_values(by=['Name']) # 'Name' 칼럼을 기준으로 오름차순으로 정렬
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

In [79]:
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False) # 'Pclass'와 'Name'을 기준으로 내림차순으로 정렬
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

### Aggregation 함수 적용

`min()`, `max()`, `sum()`, `count()`와 같은 aggregation 함수를 사용할 수 있다.
<br>
DataFrame의 경우 **DataFrame에서 바로 aggregation을 호출할 경우 모든 칼럼에 해당 aggregation을 적용**한다.

In [80]:
titanic_df.count() # 각 항목의 수 확인

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

모든 칼럼에 대해 `count()`가 수행된 것을 볼 수 있다.

In [81]:
titanic_df[['Age', 'Fare']].mean() # 'Age', 'Fare' 칼럼의 평균값을 확인

Age     29.699118
Fare    32.204208
dtype: float64

위와 같이 aggregation을 적용할 칼럼을 직접 지정할 수 있다.
<br>
이때 칼럼명이 여러개이면 대괄호('[]')를 두 개 사용하여 [[]]와 같은 식으로 표현하여야 한다는 것을 잊지 말자.

### groupby() 적용

DataFrame의 `groupby()` 사용 시 입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby된다.
<br>
그리고 **DataFrameGroupBy라는 또다른 형태의 DataFrame을 반환**한다.

In [82]:
titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


titanic_groupby의 타입이 DataFrameGroupBy인 것을 확인할 수 있다.

In [60]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

`groupby()` 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수가 적용된 것을 볼 수 있다.

In [83]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count() # 'Pclass'에 대해 groupby된 것을 선택한 칼럼들의 데이터만 나타내기
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [62]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

In [84]:
agg_format={'Age':'max', 'SibSp':'sum', 'Fare':'mean'} # 딕셔너리 형태로 만들어 aggregation을 수행
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

## 1.4.8 결손 데이터 처리하기

결손 데이터는 칼럼에 값이 없는, 즉 Null인 경우를 의미하며, 이를 넘파이의 **NaN**으로 표시한다.
<br>
기본적으로 머신러닝 알고리즘은 이 NaN값을 처리하지 않으므로 이 값을 다른 값으로 대체해야 한다.
<br>
또한 이 값은 평균, 총합 등 함수 연산 시 제외가 된다.

### isna()로 결손 데이터 여부 확인

In [85]:
titanic_df.isna().head(3) # .isna()로 결측치인지 확인

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

In [86]:
titanic_df.isna( ).sum( ) # 결측치 수의 합계 확인

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### fillna()로 missing 데이터 대체하기

In [87]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000') # .fillna()로 결측치 값 대체하기
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

주의해야 할 점은, `fillna()`를 사용할 때 DataFrame이 반환되기 때문에 DataFrame의 변수로 반환되는 데이터를 받아야 한다는 것이다.
<br>
만약 함수 자체로만 사용해 데이터 자신의 결측치를 대체하고 싶다면 **inplace=True**를 설정하면 된다.

In [88]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean()) # 'Age' 칼럼의 평균 값으로 결측치를 대체
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S') # 'S'로 'Embarked' 칼럼의 결측치를 대체
titanic_df.isna().sum() # 결측치 수를 확인하여 잘 적용되었는지 확인

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

## 1.4.9 apply lambda 식으로 데이터 가공

판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제공한다. 

In [68]:
def get_square(a):
    return a**2

print('3의 제곱은:',get_square(3))

3의 제곱은: 9


위의 기능을 아래와 같이 람다 식으로 간단하게 나타낼 수 있다.

In [69]:
lambda_square = lambda x : x ** 2
print('3의 제곱은:',lambda_square(3))

3의 제곱은: 9


위에서 'x'는 입력 인자이고, 'x**2'는 입력 인자를 기반으로 한 계산식이며, 호출 시 계산 결과가 반환된다.

lamda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우에는 보통 `map()` 함수를 결합해서 사용한다.

In [70]:
a=[1,2,3]
squares = map(lambda x : x**2, a) # 람다 함수를 리스트 a의 각 요소에 적용
list(squares)

[1, 4, 9]

In [71]:
# 새로운 칼럼을 만들고, apply() 메서드를 사용하여 'Name' 열의 각 항목에 대해 이 len 함수를 적용
titanic_df['Name_len']= titanic_df['Name'].apply(lambda x : len(x))
titanic_df[['Name','Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

조건문을 활용한 람다식도 적용할 수 있다.
<br>
이때 주의할 점은, **if 절 식보다 반환 값을 먼저 기술해야 한다**는 것이다. **lamda 식 ':' 기호의 오른편에 반환 값이 있어야 하기 때문**이다. 하지만 else 절에서는 원래대로 반환 값이 나중에 온다.

In [72]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <=15 else 'Adult' )
titanic_df[['Age','Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

In [90]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x<=15 else ('Adult' if x <= 60 else 
                                                                                  'Elderly'))
titanic_df['Age_cat'].value_counts()

Age_cat
Adult      786
Child       83
Elderly     22
Name: count, dtype: int64

In [74]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성. 
def get_category(age):
    cat = ''
    if age <= 5: cat = 'Baby'
    elif age <= 12: cat = 'Child'
    elif age <= 18: cat = 'Teenager'
    elif age <= 25: cat = 'Student'
    elif age <= 35: cat = 'Young Adult'
    elif age <= 60: cat = 'Adult'
    else : cat = 'Elderly'
    
    return cat

# lambda 식에 위에서 생성한 get_category( ) 함수를 반환값으로 지정. 
# get_category(X)는 입력값으로 ‘Age’ 컬럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age','Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult